http/tokenV2/middleware.go

```diff
@@ -0,0 +1,382 @@
+	// keys as RS512 is the most secure signature algorithm supported in
+	// the ssh-agent protocol. It seems the ssh ecosystem has generally moved
+	// beyond RSA support in favor of trying to maximize security within the
+	// scope of RSA use. Perhaps we could consider doing the same.
+	case jwa.RS512, jwa.PS512:
+		return true
+
+	// Edwards curve signatures are considered secure and are therefore supported
+	case jwa.EdDSA:
+		return true
+
+	// Explicitly reject messages signed by the "none" algorithm. This
+	// would technically be covered by the default case below but it makes
+	// the intent clear in case somebody tries to turn this from a whitelist
+	// approach into a blacklist approach in the future. Not to mention this
+	// going wrong would result in a catastrophic security hole, so it's worth
+	// having a special case for it.
+	case jwa.NoSignature:
+		return false
+
+	// Only explicitly allowed signing algorithms are acceptable
+	default:
+		return false
+	}
+}
+
+// authenticationCredential returns the token present in the Authorization
+// header of the HTTP request.
+func authenticationCredential(context echo.Context) string {
+	// Extract the authorization header from the request
+	credential := context.Request().Header.Get("Authorization")
+	if credential == "" {
+		return ""
+	}
+
+	// Split the authorization header on whitespace
+	fields := strings.Fields(credential)
+
+	// Ignore any header unless it has precisely 2 fields
+	if len(fields) != 2 {
+		return ""
+	}
+
+	// Ignore non-bearer headers
+	if strings.ToLower(fields[0]) != "bearer" {
+		return ""
+	}
+
+	// Return the supplied credential
+	return fields[1]
+}
+
+// unauthorizedError returns an echo unauthorized error
+func unauthorizedError(context echo.Context, reason error) *echo.HTTPError {
+	// Explicitly set the response to 401 Unauthorized rather than relying on any default behaviour
+	context.String(http.StatusUnauthorized, http.StatusText(http.StatusUnauthorized))
+
+	// Set an empty username for this context
+	context.Set(core.UserContextKey, "")
+
+	// Log an entry in the audit log about this failure
+	auditLog(context, defaultActor(context), audit.AccessDeniedEvent)
+
+	// Return the appropriate echo error to ensure complete logging
+	return &echo.HTTPError{
+		Code:     http.StatusUnauthorized,
+		Message:  http.StatusText(http.StatusUnauthorized),
+		Internal: reason,
+	}
+}
+
+// defaultActor returns some suitable default representation for the remote actor
+func defaultActor(context echo.Context) string {
+	// If available use the real IP of the caller as the default actor name
+	realIP := context.RealIP()
+	if realIP != "" {
+		return realIP
+	}
+
+	return "unknown"
+}
+
+// auditLog logs a security event about an actor given a certain echo context
+func auditLog(context echo.Context, actor string, event string) {
+	auditContext := audit.Context(context.Request().Context(), actor, "tokenV2", "middleware")
+	audit.Log(auditContext, log.Logger(), event)
```

Comment:
this doesn't actually log, you still need to call a logger method
```suggestion
	audit.Log(auditContext, log.Logger(), event).Info("Some event description")
```

Add assertion for audit log to the tests (`audit.CaptureLogs`, `audit.AssertContains` or something, check other tests)